ortools/linear_solver/xpress_interface.cc

```diff
@@ -1647,6 +1698,21 @@ MPSolver::ResultStatus XpressInterface::Solve(MPSolverParameters const& param) {
   // Set the hint (if any)
   this->AddSolutionHintToOptimizer();
 
+  // Add opt node callback to optimizer. We have to do this here (just before
+  // solve) to make sure the variables are fully initialized
+  std::unique_ptr<XpressMPCallbackContext> xpress_context;
+  MPCallbackWithXpressContext mp_callback_with_context;
+  if (callback_ != nullptr) {
+    xpress_context = std::make_unique<XpressMPCallbackContext>(
+        mLp, mMip, solver_->objective_->maximization(),
+        solver_->NumVariables());
+    mp_callback_with_context.callback = callback_;
+    mp_callback_with_context.context = xpress_context.get();
+    CHECK_STATUS(XPRSaddcbintsol(mLp, XpressOptNodeCallbackImpl,
```

Comment:
This is now the correct callback to use for the intended behavior.
But now we have a different problem: You can no longer call `SuggestSolution` from the callback :-(
This time the problem is on the Xpress side: At the moment you cannot call XPRSaddmipsol() from an intsol callback. We are working on supporting this but it is not clear when this will end up in the product.
So either you go back to optnode callback or you disable the `SuggestSolution` thing for the time being.
Note that using the optnode callback has another problem: If multiple heuristics solutions are found between two nodes then you may miss some of them since there is no optnode callback invocation between them.
So at the moment we are a bit limited by what Xpress can actually do :-(

Reply:
okay it's not a big problem for us, we don't need this feature actually. I just implemented it because we had already implemented "SetHint" from outside the callback, so the extra effort was negligible. 
I will disable calling XPRSaddmipsol when the call back "XPRSaddcbintsol" is called for now (I can detect this using the context's Event()==kMipSolution condition), and add a comment so we can re-activate it later.
thanks for the heads up!